File: MPI/count_sort/count_sort_seq.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define N 90000
#define UPPER N*4
#define LOWER 1

void rand_init_array(int array[], int n, int upper, int lower);
void display_array(int array[], int n);
void count_sort(int a[], int n);
void display_time(clock_t start, clock_t end);

int main(void){ 

    int array[N];

    rand_init_array(array, N, UPPER, LOWER);

    // (void) printf("Initial array: ");
    // display_array(array, N);

    (void) printf("Sorting began...\n\n");
    double begin = clock();
    count_sort(array, N);
    double end = clock();

    display_time(begin, end);
    
    // (void) printf("\n\nSorted array: ");
    // display_array(array, N);

    return 0;

}

/*
 * Function:  rand_init_array 
 * --------------------
 * Fills an integer array with random numbers
 *
 *  array: the array that will be filled with numbers
 *  n: number of elements in the array
 *  upper: highest value of random number
 *  lower: lowest value of random number
 *
 */

void rand_init_array(int array[], int n, int upper, int lower){
    int i;    
    for (i=0; i<n; ++i)
        array[i] = (rand() % (upper - lower + 1)) + lower;
}

/*
 * Function:  display_array 
 * --------------------
 * Prints an integer array to user
 *
 *  array: the array that will be printed
 *  n: number of elements in the array
 *
 */

void display_array(int array[], int n){
    (void) printf("[ ");
    int i;
    for (i=0; i<n; ++i)
        (void) printf("%d ", array[i]);
    (void) printf("]\n\n");
}

/*
 * Function:  count_sort 
 * --------------------
 * Sorts an integer array, using the count sort algorithm (enumeration sort)
 *
 *  a: the array that will be sorted
 *  n: number of elements in the array
 *
 */

void count_sort(int a[], int n) {
    int i, j, count;
    int* temp = malloc(n*sizeof(int));
    for (i = 0; i < n; ++i) {
        count = 0;
        for (j = 0; j < n; ++j)
            if (a[j] < a[i])
                count++;
            else if (a[j] == a[i] && j < i)
                count++;
        temp[count] = a[i];
    }
    
    memcpy(a, temp, n*sizeof(int));
    free(temp);
}

/*
 * Function:  display_time 
 * --------------------
 * Prints the time (seconds) elapsed for sorting
 *
 *  start: the time (seconds) in which the sorting started
 *  end: the time (seconds) in which the sorting finished
 *
 */

void display_time(clock_t start, clock_t end){
    (void) printf("Time spent for sorting: %g seconds\n", (double)(end-start) / CLOCKS_PER_SEC);
}

File: MPI/mpi4py/count_sort/count_sort_seq.py

import numpy as np
from mpi4py import MPI

N = 10
LOWER = 0
UPPER = N*4

def count_sort(a, n):
    '''
    Function:  count_sort 
    --------------------
    Sorts an integer array, using the count sort algorithm (enumeration sort)

    a: the array that will be sorted
    n: number of elements in the array=
    '''
    temp = np.empty(dtype=int, shape=n)
    for i in range(n):
        count = 0
        for j in range(n):
            if a[j] < a[i]:
                count += 1
            elif a[j] == a[i] and j < i:
                count += 1
        temp[count] = a[i]

    a[:] = temp

if __name__ == '__main__':
    a = np.random.randint(low=LOWER, high=UPPER, size=N)
    print(f'\nInitial Array: {a}')
    begin = MPI.Wtime()
    count_sort(a, N)
    end = MPI.Wtime()
    print(f'\nSorted Array: {a}')
    print(f'\nTime elapsed for operation: {end-begin} seconds.\n')
